superset/security/manager.py

```diff
@@ -987,15 +988,24 @@ def raise_for_access(  # pylint: disable=too-many-arguments,too-many-branches
                 )
 
         if datasource or query_context or viz:
+            extra_jwt = None
             if query_context:
                 datasource = query_context.datasource
+                extra_jwt = query_context.extra_jwt
             elif viz:
                 datasource = viz.datasource
+                extra_jwt = viz.extra_jwt
 
             assert datasource
 
+            dashboard_data_context = dashboard_jwt_manager.parse_jwt(extra_jwt)
+
+            data_source_allowed_in_dashboard = (
+                datasource.id in dashboard_data_context.dataset_ids
+            )
             if not (
-                self.can_access_schema(datasource)
+                data_source_allowed_in_dashboard
+                or self.can_access_schema(datasource)
```

Comment:
I'd rewrite this to make sure there's no way the new logic can leak into the current flow without the FF. Something like
```python
data_source_allowed_in_dashboard = False
if feature_flag_manager.is_feature_enabled("DASHBOARD_RBAC"):
    dashboard_data_context = dashboard_jwt_manager.parse_jwt(extra_jwt)
    data_source_allowed_in_dashboard = (
        datasource.id in dashboard_data_context.dataset_ids
    )
```